presto-google-sheets/src/main/java/com/facebook/presto/google/sheets/SheetsClient.java

```diff
@@ -0,0 +1,227 @@
+import static com.facebook.presto.google.sheets.SheetsErrorCode.SHEETS_TABLE_LOAD_ERROR;
+import static com.facebook.presto.google.sheets.SheetsErrorCode.SHEETS_UNKNOWN_TABLE_ERROR;
+import static com.google.api.client.googleapis.javanet.GoogleNetHttpTransport.newTrustedTransport;
+import static com.google.common.cache.CacheLoader.from;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class SheetsClient
+{
+    private static final Logger log = Logger.get(SheetsClient.class);
+
+    private static final String APPLICATION_NAME = "Presto google sheets integration";
+    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+    private static final List<String> SCOPES = ImmutableList.of(SheetsScopes.SPREADSHEETS_READONLY);
+
+    private final LoadingCache<String, Optional<String>> tableSheetMappingCache;
+    private final LoadingCache<String, List<List<Object>>> sheetDataCache;
+
+    private final String metadataSheetId;
+    private final String credentialsFilePath;
+
+    private final Sheets sheetsService;
+
+    @Inject
+    public SheetsClient(SheetsConfig config, JsonCodec<Map<String, List<SheetsTable>>> catalogCodec)
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(catalogCodec, "catalogCodec is null");
+
+        this.metadataSheetId = config.getMetadataSheetId();
+        this.credentialsFilePath = config.getCredentialsFilePath();
+
+        try {
+            this.sheetsService = new Sheets.Builder(newTrustedTransport(), JSON_FACTORY, getCredentials()).setApplicationName(APPLICATION_NAME).build();
+        }
+        catch (GeneralSecurityException | IOException e) {
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        long expiresAfterWriteMillis = config.getSheetsDataExpireAfterWrite().toMillis();
+        long maxCacheSize = config.getSheetsDataMaxCacheSize();
+
+        this.tableSheetMappingCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize)
+                .build(new CacheLoader<String, Optional<String>>()
+                {
+                    @Override
+                    public Optional<String> load(String tableName)
+                    {
+                        return getSheetExpressionForTable(tableName);
+                    }
+                });
+
+        this.sheetDataCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize).build(from(this::readAllValuesFromSheetExpression));
+    }
+
+    public Optional<SheetsTable> getTable(String tableName)
+    {
+        List<List<Object>> values = readAllValues(tableName);
+        if (values.size() > 0) {
+            ImmutableList.Builder<SheetsColumn> columns = ImmutableList.builder();
+            Set<String> columnNames = new HashSet<>();
+            // Assuming 1st line is always header
+            List<Object> header = values.get(0);
```

Comment:
Just wondering what if the first line is not header? May we document this requirement in the `presto-docs`?

Reply:
Yes, current implementation assumes first row is the header, and good point about doc. I will definitely update the docs. Thanks.